{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5b557744-a55b-4a9a-8343-564e82124e2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "proj_tumorvsctrl <- data.frame(tcga=c(\"tcga_brca\", \"tcga_coad\",\"tcga_esca\",\"tcga_lgg\",\"tcga_luad\",\"tcga_lusc\",\"tcga_paad\",\"tcga_skcm\",\"tcga_stad\"),\n",
    "                          gtex=c(\"Breast\",\"Colon\",\"Esophagus\",\"Brain\",\"Lung\",\"Lung\",\"Pancreas\",\"Skin\",\"Stomach\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0d691ab9-46d3-4fe0-b039-90763d49198b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(x):\n",
      "“'breast_tcga_brca' already exists”\n",
      "Warning message in dir.create(x):\n",
      "“'colon_tcga_coad' already exists”\n",
      "Warning message in dir.create(x):\n",
      "“'esophagus_tcga_esca' already exists”\n",
      "Warning message in dir.create(x):\n",
      "“'brain_tcga_lgg' already exists”\n",
      "Warning message in dir.create(x):\n",
      "“'lung_tcga_luad' already exists”\n",
      "Warning message in dir.create(x):\n",
      "“'lung_tcga_lusc' already exists”\n",
      "Warning message in dir.create(x):\n",
      "“'pancreas_tcga_paad' already exists”\n",
      "Warning message in dir.create(x):\n",
      "“'skin_tcga_skcm' already exists”\n",
      "Warning message in dir.create(x):\n",
      "“'stomach_tcga_stad' already exists”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<ol>\n",
       "\t<li>FALSE</li>\n",
       "\t<li>FALSE</li>\n",
       "\t<li>FALSE</li>\n",
       "\t<li>FALSE</li>\n",
       "\t<li>FALSE</li>\n",
       "\t<li>FALSE</li>\n",
       "\t<li>FALSE</li>\n",
       "\t<li>FALSE</li>\n",
       "\t<li>FALSE</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate}\n",
       "\\item FALSE\n",
       "\\item FALSE\n",
       "\\item FALSE\n",
       "\\item FALSE\n",
       "\\item FALSE\n",
       "\\item FALSE\n",
       "\\item FALSE\n",
       "\\item FALSE\n",
       "\\item FALSE\n",
       "\\end{enumerate}\n"
      ],
      "text/markdown": [
       "1. FALSE\n",
       "2. FALSE\n",
       "3. FALSE\n",
       "4. FALSE\n",
       "5. FALSE\n",
       "6. FALSE\n",
       "7. FALSE\n",
       "8. FALSE\n",
       "9. FALSE\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[[1]]\n",
       "[1] FALSE\n",
       "\n",
       "[[2]]\n",
       "[1] FALSE\n",
       "\n",
       "[[3]]\n",
       "[1] FALSE\n",
       "\n",
       "[[4]]\n",
       "[1] FALSE\n",
       "\n",
       "[[5]]\n",
       "[1] FALSE\n",
       "\n",
       "[[6]]\n",
       "[1] FALSE\n",
       "\n",
       "[[7]]\n",
       "[1] FALSE\n",
       "\n",
       "[[8]]\n",
       "[1] FALSE\n",
       "\n",
       "[[9]]\n",
       "[1] FALSE\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "lapply(tolower(paste(proj_tumorvsctrl$gtex, proj_tumorvsctrl$tcga, sep=\"_\")), FUN=function(x) dir.create(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a3d1cc2c-de8c-42a6-8f07-4fe0a57a5597",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(paste(\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/\", :\n",
      "“'/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/breast_tcga_brca' already exists”\n",
      "Warning message in dir.create(paste(\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/\", :\n",
      "“'/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/colon_tcga_coad' already exists”\n",
      "Warning message in dir.create(paste(\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/\", :\n",
      "“'/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/esophagus_tcga_esca' already exists”\n",
      "Warning message in dir.create(paste(\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/\", :\n",
      "“'/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/brain_tcga_lgg' already exists”\n",
      "Warning message in dir.create(paste(\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/\", :\n",
      "“'/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/lung_tcga_luad' already exists”\n",
      "Warning message in dir.create(paste(\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/\", :\n",
      "“'/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/lung_tcga_lusc' already exists”\n",
      "Warning message in dir.create(paste(\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/\", :\n",
      "“'/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/pancreas_tcga_paad' already exists”\n",
      "Warning message in dir.create(paste(\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/\", :\n",
      "“'/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/skin_tcga_skcm' already exists”\n",
      "Warning message in dir.create(paste(\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/\", :\n",
      "“'/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/stomach_tcga_stad' already exists”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<ol>\n",
       "\t<li>FALSE</li>\n",
       "\t<li>FALSE</li>\n",
       "\t<li>FALSE</li>\n",
       "\t<li>FALSE</li>\n",
       "\t<li>FALSE</li>\n",
       "\t<li>FALSE</li>\n",
       "\t<li>FALSE</li>\n",
       "\t<li>FALSE</li>\n",
       "\t<li>FALSE</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate}\n",
       "\\item FALSE\n",
       "\\item FALSE\n",
       "\\item FALSE\n",
       "\\item FALSE\n",
       "\\item FALSE\n",
       "\\item FALSE\n",
       "\\item FALSE\n",
       "\\item FALSE\n",
       "\\item FALSE\n",
       "\\end{enumerate}\n"
      ],
      "text/markdown": [
       "1. FALSE\n",
       "2. FALSE\n",
       "3. FALSE\n",
       "4. FALSE\n",
       "5. FALSE\n",
       "6. FALSE\n",
       "7. FALSE\n",
       "8. FALSE\n",
       "9. FALSE\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[[1]]\n",
       "[1] FALSE\n",
       "\n",
       "[[2]]\n",
       "[1] FALSE\n",
       "\n",
       "[[3]]\n",
       "[1] FALSE\n",
       "\n",
       "[[4]]\n",
       "[1] FALSE\n",
       "\n",
       "[[5]]\n",
       "[1] FALSE\n",
       "\n",
       "[[6]]\n",
       "[1] FALSE\n",
       "\n",
       "[[7]]\n",
       "[1] FALSE\n",
       "\n",
       "[[8]]\n",
       "[1] FALSE\n",
       "\n",
       "[[9]]\n",
       "[1] FALSE\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(paste(\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/\", :\n",
      "“'/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/breast_tcga_brca/breast_tcga_brca_male' already exists”\n",
      "Warning message in dir.create(paste(\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/\", :\n",
      "“'/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/colon_tcga_coad/colon_tcga_coad_male' already exists”\n",
      "Warning message in dir.create(paste(\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/\", :\n",
      "“'/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/esophagus_tcga_esca/esophagus_tcga_esca_male' already exists”\n",
      "Warning message in dir.create(paste(\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/\", :\n",
      "“'/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/brain_tcga_lgg/brain_tcga_lgg_male' already exists”\n",
      "Warning message in dir.create(paste(\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/\", :\n",
      "“'/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/lung_tcga_luad/lung_tcga_luad_male' already exists”\n",
      "Warning message in dir.create(paste(\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/\", :\n",
      "“'/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/lung_tcga_lusc/lung_tcga_lusc_male' already exists”\n",
      "Warning message in dir.create(paste(\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/\", :\n",
      "“'/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/pancreas_tcga_paad/pancreas_tcga_paad_male' already exists”\n",
      "Warning message in dir.create(paste(\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/\", :\n",
      "“'/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/skin_tcga_skcm/skin_tcga_skcm_male' already exists”\n",
      "Warning message in dir.create(paste(\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/\", :\n",
      "“'/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/stomach_tcga_stad/stomach_tcga_stad_male' already exists”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<ol>\n",
       "\t<li>FALSE</li>\n",
       "\t<li>FALSE</li>\n",
       "\t<li>FALSE</li>\n",
       "\t<li>FALSE</li>\n",
       "\t<li>FALSE</li>\n",
       "\t<li>FALSE</li>\n",
       "\t<li>FALSE</li>\n",
       "\t<li>FALSE</li>\n",
       "\t<li>FALSE</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate}\n",
       "\\item FALSE\n",
       "\\item FALSE\n",
       "\\item FALSE\n",
       "\\item FALSE\n",
       "\\item FALSE\n",
       "\\item FALSE\n",
       "\\item FALSE\n",
       "\\item FALSE\n",
       "\\item FALSE\n",
       "\\end{enumerate}\n"
      ],
      "text/markdown": [
       "1. FALSE\n",
       "2. FALSE\n",
       "3. FALSE\n",
       "4. FALSE\n",
       "5. FALSE\n",
       "6. FALSE\n",
       "7. FALSE\n",
       "8. FALSE\n",
       "9. FALSE\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[[1]]\n",
       "[1] FALSE\n",
       "\n",
       "[[2]]\n",
       "[1] FALSE\n",
       "\n",
       "[[3]]\n",
       "[1] FALSE\n",
       "\n",
       "[[4]]\n",
       "[1] FALSE\n",
       "\n",
       "[[5]]\n",
       "[1] FALSE\n",
       "\n",
       "[[6]]\n",
       "[1] FALSE\n",
       "\n",
       "[[7]]\n",
       "[1] FALSE\n",
       "\n",
       "[[8]]\n",
       "[1] FALSE\n",
       "\n",
       "[[9]]\n",
       "[1] FALSE\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(paste(\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/\", :\n",
      "“'/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/breast_tcga_brca/breast_tcga_brca_female' already exists”\n",
      "Warning message in dir.create(paste(\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/\", :\n",
      "“'/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/colon_tcga_coad/colon_tcga_coad_female' already exists”\n",
      "Warning message in dir.create(paste(\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/\", :\n",
      "“'/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/esophagus_tcga_esca/esophagus_tcga_esca_female' already exists”\n",
      "Warning message in dir.create(paste(\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/\", :\n",
      "“'/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/brain_tcga_lgg/brain_tcga_lgg_female' already exists”\n",
      "Warning message in dir.create(paste(\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/\", :\n",
      "“'/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/lung_tcga_luad/lung_tcga_luad_female' already exists”\n",
      "Warning message in dir.create(paste(\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/\", :\n",
      "“'/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/lung_tcga_lusc/lung_tcga_lusc_female' already exists”\n",
      "Warning message in dir.create(paste(\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/\", :\n",
      "“'/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/pancreas_tcga_paad/pancreas_tcga_paad_female' already exists”\n",
      "Warning message in dir.create(paste(\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/\", :\n",
      "“'/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/skin_tcga_skcm/skin_tcga_skcm_female' already exists”\n",
      "Warning message in dir.create(paste(\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/\", :\n",
      "“'/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/stomach_tcga_stad/stomach_tcga_stad_female' already exists”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<ol>\n",
       "\t<li>FALSE</li>\n",
       "\t<li>FALSE</li>\n",
       "\t<li>FALSE</li>\n",
       "\t<li>FALSE</li>\n",
       "\t<li>FALSE</li>\n",
       "\t<li>FALSE</li>\n",
       "\t<li>FALSE</li>\n",
       "\t<li>FALSE</li>\n",
       "\t<li>FALSE</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate}\n",
       "\\item FALSE\n",
       "\\item FALSE\n",
       "\\item FALSE\n",
       "\\item FALSE\n",
       "\\item FALSE\n",
       "\\item FALSE\n",
       "\\item FALSE\n",
       "\\item FALSE\n",
       "\\item FALSE\n",
       "\\end{enumerate}\n"
      ],
      "text/markdown": [
       "1. FALSE\n",
       "2. FALSE\n",
       "3. FALSE\n",
       "4. FALSE\n",
       "5. FALSE\n",
       "6. FALSE\n",
       "7. FALSE\n",
       "8. FALSE\n",
       "9. FALSE\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[[1]]\n",
       "[1] FALSE\n",
       "\n",
       "[[2]]\n",
       "[1] FALSE\n",
       "\n",
       "[[3]]\n",
       "[1] FALSE\n",
       "\n",
       "[[4]]\n",
       "[1] FALSE\n",
       "\n",
       "[[5]]\n",
       "[1] FALSE\n",
       "\n",
       "[[6]]\n",
       "[1] FALSE\n",
       "\n",
       "[[7]]\n",
       "[1] FALSE\n",
       "\n",
       "[[8]]\n",
       "[1] FALSE\n",
       "\n",
       "[[9]]\n",
       "[1] FALSE\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "lapply(tolower(paste(proj_tumorvsctrl$gtex, proj_tumorvsctrl$tcga, sep=\"_\")), FUN=function(x)\n",
    "dir.create(paste(\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/\",x,\"/\",sep=\"\")))\n",
    "lapply(tolower(paste(proj_tumorvsctrl$gtex, proj_tumorvsctrl$tcga, sep=\"_\")), FUN=function(x)\n",
    "dir.create(paste(\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/\",x,\"/\",x,\"_male\",sep=\"\")))\n",
    "lapply(tolower(paste(proj_tumorvsctrl$gtex, proj_tumorvsctrl$tcga, sep=\"_\")), FUN=function(x)\n",
    "dir.create(paste(\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/\",x,\"/\",x,\"_female\",sep=\"\")))              "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "9ee054f9-f4a9-489e-bcdf-b542950c8e47",
   "metadata": {},
   "outputs": [],
   "source": [
    "creat_param_yml_male <- function(project){\n",
    "params_custom_circuits <- \n",
    "paste(\"# output folder\",\"\\n\",\n",
    "\"output_folder: \\\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/\", tolower(paste(proj_tumorvsctrl[proj_tumorvsctrl$tcga == project,\"gtex\"],project,sep=\"_\")),\"/\",tolower(paste(proj_tumorvsctrl[proj_tumorvsctrl$tcga == project,\"gtex\"],project,sep=\"_\")),\"_male\\\"\",\"\\n\",\n",
    "    \n",
    "\"#Drugbank compressed XML file\",\"\\n\",\n",
    "\"drugbank: \\\"https://zenodo.org/records/7969017/files/drugbank-v050108.xml.gz\\\"\",\"\\n\",\n",
    "\n",
    "\"# The Anatomical Therapeutic Chemical (ATC) Classification table\",\"\\n\",\n",
    "      \n",
    "\"atc: \\\"https://zenodo.org/records/7969017/files/ATC.csv.gz\\\"\",\"\\n\",\n",
    "\"# HiPathia's list of physiological KEGG signaling pathways with GO/Uniprot functional annotations\",\"\\n\",\n",
    "\"phys_paths: \\\"https://zenodo.org/records/7969017/files/physiological_paths.tsv\\\"\",\"\\n\",\n",
    "\n",
    "\"# The Genotype-Tissue Expression (GTEx) RNA-Seq gene read counts\",\"\\n\",\n",
    "\"gtex: \\\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_5_genexpresion_tcga/\",\n",
    "proj_tumorvsctrl[proj_tumorvsctrl$tcga == project,\"gtex\"],\n",
    "\"_geneexp_male.tsv\\\"\",\"\\n\",\n",
    "\n",
    "\"# Hipathia disease map circuit file. Uncomment to use a custom disease map list of circuits\",\"\\n\",\n",
    "\"disease_map_circuits: \\\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_2_circuits_sign_limma_tumvsctrl/circuits_male_\",\n",
    "      project,\".tsv\\\"\",\"\\n\",\n",
    "\n",
    "\"# Uncomment the following line to feed drexml with a custom gene target list (skip drugbank preprocess)\",\"\\n\",\n",
    "\n",
    "\"drexml_target_genes: \\\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_3_drugbank_used/09_3_drugbank_gene_targets_\",\n",
    "      project,\".tsv\\\"\",\"\\n\"\n",
    "\n",
    "      ,sep=\"\")\n",
    "params_custom_circuits_file <- paste(\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/code/09_drug_repository/09_4_drexml_gtex/\",tolower(paste(proj_tumorvsctrl[proj_tumorvsctrl$tcga == project,\"gtex\"],project,sep=\"_\")),\"/params_custom_circuits_male.yml\",sep=\"\")\n",
    "writeLines(params_custom_circuits,params_custom_circuits_file)\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "34661ab8-c12e-430c-a58f-91c57be39838",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate}\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\end{enumerate}\n"
      ],
      "text/markdown": [
       "1. NULL\n",
       "2. NULL\n",
       "3. NULL\n",
       "4. NULL\n",
       "5. NULL\n",
       "6. NULL\n",
       "7. NULL\n",
       "8. NULL\n",
       "9. NULL\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[[1]]\n",
       "NULL\n",
       "\n",
       "[[2]]\n",
       "NULL\n",
       "\n",
       "[[3]]\n",
       "NULL\n",
       "\n",
       "[[4]]\n",
       "NULL\n",
       "\n",
       "[[5]]\n",
       "NULL\n",
       "\n",
       "[[6]]\n",
       "NULL\n",
       "\n",
       "[[7]]\n",
       "NULL\n",
       "\n",
       "[[8]]\n",
       "NULL\n",
       "\n",
       "[[9]]\n",
       "NULL\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "lapply(proj_tumorvsctrl$tcga, FUN=function(x) creat_param_yml_male(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "eb945d07-362e-4562-b6b0-b6296523d3ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "creat_param_yml_female <- function(project){\n",
    "params_custom_circuits <- \n",
    "paste(\"# output folder\",\"\\n\",\n",
    "\"output_folder: \\\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_4_drexml_gtex/\", tolower(paste(proj_tumorvsctrl[proj_tumorvsctrl$tcga == project,\"gtex\"],project,sep=\"_\")),\"/\",tolower(paste(proj_tumorvsctrl[proj_tumorvsctrl$tcga == project,\"gtex\"],project,sep=\"_\")),\"_female\\\"\",\"\\n\",\n",
    "\n",
    "\"#Drugbank compressed XML file\",\"\\n\",\n",
    "\"drugbank: \\\"https://zenodo.org/records/7969017/files/drugbank-v050108.xml.gz\\\"\",\"\\n\",\n",
    "\n",
    "\"# The Anatomical Therapeutic Chemical (ATC) Classification table\",\"\\n\",\n",
    "      \n",
    "\"atc: \\\"https://zenodo.org/records/7969017/files/ATC.csv.gz\\\"\",\"\\n\",\n",
    "\"# HiPathia's list of physiological KEGG signaling pathways with GO/Uniprot functional annotations\",\"\\n\",\n",
    "\"phys_paths: \\\"https://zenodo.org/records/7969017/files/physiological_paths.tsv\\\"\",\"\\n\",\n",
    "\n",
    "\"# The Genotype-Tissue Expression (GTEx) RNA-Seq gene read counts\",\"\\n\",\n",
    "\"gtex: \\\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_5_genexpresion_tcga/\",\n",
    "proj_tumorvsctrl[proj_tumorvsctrl$tcga == project,\"gtex\"],\n",
    "\"_geneexp_female.tsv\\\"\",\"\\n\",\n",
    "\n",
    "\"# Hipathia disease map circuit file. Uncomment to use a custom disease map list of circuits\",\"\\n\",\n",
    "\"disease_map_circuits: \\\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_2_circuits_sign_limma_tumvsctrl/circuits_female_\",\n",
    "      project,\".tsv\\\"\",\"\\n\",\n",
    "\n",
    "\"# Uncomment the following line to feed drexml with a custom gene target list (skip drugbank preprocess)\",\"\\n\",\n",
    "\n",
    "\"drexml_target_genes: \\\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/results/09_drug_repository/09_3_drugbank_used/09_3_drugbank_gene_targets_\",\n",
    "      project,\".tsv\\\"\",\"\\n\"\n",
    "\n",
    "      ,sep=\"\")\n",
    "params_custom_circuits_file <- paste(\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/code/09_drug_repository/09_4_drexml_gtex/\",tolower(paste(proj_tumorvsctrl[proj_tumorvsctrl$tcga == project,\"gtex\"],project,sep=\"_\")),\"/params_custom_circuits_female.yml\",sep=\"\")\n",
    "writeLines(params_custom_circuits,params_custom_circuits_file)\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "3e38d488-3f76-415f-b24f-25628e841d92",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate}\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\end{enumerate}\n"
      ],
      "text/markdown": [
       "1. NULL\n",
       "2. NULL\n",
       "3. NULL\n",
       "4. NULL\n",
       "5. NULL\n",
       "6. NULL\n",
       "7. NULL\n",
       "8. NULL\n",
       "9. NULL\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[[1]]\n",
       "NULL\n",
       "\n",
       "[[2]]\n",
       "NULL\n",
       "\n",
       "[[3]]\n",
       "NULL\n",
       "\n",
       "[[4]]\n",
       "NULL\n",
       "\n",
       "[[5]]\n",
       "NULL\n",
       "\n",
       "[[6]]\n",
       "NULL\n",
       "\n",
       "[[7]]\n",
       "NULL\n",
       "\n",
       "[[8]]\n",
       "NULL\n",
       "\n",
       "[[9]]\n",
       "NULL\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "lapply(proj_tumorvsctrl$tcga, FUN=function(x) creat_param_yml_female(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "77369f7b-4d2a-4697-9bb3-e4fbf79aaf65",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate}\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\end{enumerate}\n"
      ],
      "text/markdown": [
       "1. NULL\n",
       "2. NULL\n",
       "3. NULL\n",
       "4. NULL\n",
       "5. NULL\n",
       "6. NULL\n",
       "7. NULL\n",
       "8. NULL\n",
       "9. NULL\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[[1]]\n",
       "NULL\n",
       "\n",
       "[[2]]\n",
       "NULL\n",
       "\n",
       "[[3]]\n",
       "NULL\n",
       "\n",
       "[[4]]\n",
       "NULL\n",
       "\n",
       "[[5]]\n",
       "NULL\n",
       "\n",
       "[[6]]\n",
       "NULL\n",
       "\n",
       "[[7]]\n",
       "NULL\n",
       "\n",
       "[[8]]\n",
       "NULL\n",
       "\n",
       "[[9]]\n",
       "NULL\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "create_file_launcher_custom_circuits_male <- function(project) {\n",
    "  filename <- paste0(\n",
    "    \"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/code/09_drug_repository/09_4_drexml_gtex/\",\n",
    "    tolower(paste(proj_tumorvsctrl[proj_tumorvsctrl$tcga == project,\"gtex\"],project,sep=\"_\")),\n",
    "    \"/launcher_custom_circuits_male.sh\"\n",
    "  )\n",
    "  \n",
    "  script <- paste(\n",
    "    \"#!/bin/bash\\n\",\n",
    "    \"#SBATCH -J dr-schizo-launcher\\n\",\n",
    "    \"#SBATCH -t 0\\n\",\n",
    "    \"#SBATCH --mem 8G\\n\",\n",
    "    \"#SBATCH -n 1\\n\",\n",
    "    \"#SBATCH -o pipeline_custom_circuits.log\\n\\n\",\n",
    "    \"###################################################\\n\",\n",
    "    \"# SET THE FOLLOWING VARIABLES TO MATCH YOUR NEEDS #\\n\",\n",
    "    \"###################################################\\n\",\n",
    "    \"SINGULARITY_MOD='Singularity/3.9.7'\\n\",\n",
    "    \"MAMBA_MOD='Mamba/4.14.0-0'\\n\",\n",
    "    \"PARAMS_FILE='/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/code/09_drug_repository/09_4_drexml_gtex/\",\n",
    "    tolower(paste(proj_tumorvsctrl[proj_tumorvsctrl$tcga == project,\"gtex\"],project,sep=\"_\")),\n",
    "    \"/params_custom_circuits_male.yml'\\n\",\n",
    "    \"CONDA_INIT_SCRIPT='/mnt/lustre/expanse/software/easybuild/Rocky/8.5/Skylake/software/Mamba/4.14.0-0/etc/profile.d/conda.sh'\\n\",\n",
    "    \"CONDA_ENV='nextflow'\\n\",\n",
    "    \"UPDATE_NEXTFLOW='TRUE'\\n\\n\",\n",
    "    \"# The PIPELINE_REPO variable must point to the github repository (user/project) or the main.nf full path (in your local -cloned- repository)\\n\",\n",
    "    \"#PIPELINE_REPO='babelomics/dr-pipeline'\\n\",\n",
    "    \"PIPELINE_REPO='/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/code/09_drug_repository/dr-pipeline/main.nf'\\n\\n\",\n",
    "    \"########\\n\",\n",
    "    \"# INIT #\\n\",\n",
    "    \"########\\n\",\n",
    "    \"# check params file\\n\",\n",
    "    \"[[ -f \\\"${PARAMS_FILE}\\\" ]] || { echo \\\"Params file ${PARAMS_FILE} not found!\\\" && exit 1; }\\n\\n\",\n",
    "    \"# check if singularity module is loaded\\n\",\n",
    "    \"[[ $(module is-loaded ${SINGULARITY_MOD}) ]] || module load ${SINGULARITY_MOD}\\n\\n\",\n",
    "    \"# check if mamba module is loaded\\n\",\n",
    "    \"[[ $(module is-loaded ${MAMBA_MOD}) ]] || module load ${MAMBA_MOD}\\n\\n\",\n",
    "    \"# initialize conda\\n\",\n",
    "    \"source ${CONDA_INIT_SCRIPT} || { echo \\\"The conda environment couldn't be initialized!\\\" && exit 1; }\\n\\n\",\n",
    "    \"# check conda env\\n\",\n",
    "    \"if [[ ! $(conda info --envs |grep ${CONDA_ENV}) ]]\\n\",\n",
    "    \"then\\n\",\n",
    "    \"  echo \\\"${CONDA_ENV} environment not found. Trying to create a new conda environment...\\\"\\n\",\n",
    "    \"  conda create -y -n ${CONDA_ENV} bioconda::nextflow\\n\",\n",
    "    \"fi\\n\\n\",\n",
    "    \"# activate the conda environment\\n\",\n",
    "    \"conda activate ${CONDA_ENV} || { echo \\\"The conda environment couldn't be activated!\\\" && exit 1; }\\n\\n\",\n",
    "    \"# update nextflow\\n\",\n",
    "    \"[[ \\\"${UPDATE_NEXTFLOW}\\\" ]] && nextflow self-update\\n\\n\",\n",
    "    \"########\\n\",\n",
    "    \"# MAIN #\\n\",\n",
    "    \"########\\n\",\n",
    "    \"# run the workflow from the github repository\\n\",\n",
    "    \"nextflow run ${PIPELINE_REPO} -profile slurm,singularity -resume -params-file ${PARAMS_FILE}\\n\",\n",
    "    sep=\"\"\n",
    "  )\n",
    "  \n",
    "  write(script, file = filename)\n",
    "}\n",
    "\n",
    "\n",
    "lapply(proj_tumorvsctrl$tcga, FUN=function(x) create_file_launcher_custom_circuits_male(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "a41febf1-3c8d-439a-b24a-6ba67d271ad9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate}\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\end{enumerate}\n"
      ],
      "text/markdown": [
       "1. NULL\n",
       "2. NULL\n",
       "3. NULL\n",
       "4. NULL\n",
       "5. NULL\n",
       "6. NULL\n",
       "7. NULL\n",
       "8. NULL\n",
       "9. NULL\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[[1]]\n",
       "NULL\n",
       "\n",
       "[[2]]\n",
       "NULL\n",
       "\n",
       "[[3]]\n",
       "NULL\n",
       "\n",
       "[[4]]\n",
       "NULL\n",
       "\n",
       "[[5]]\n",
       "NULL\n",
       "\n",
       "[[6]]\n",
       "NULL\n",
       "\n",
       "[[7]]\n",
       "NULL\n",
       "\n",
       "[[8]]\n",
       "NULL\n",
       "\n",
       "[[9]]\n",
       "NULL\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "create_file_launcher_custom_circuits_female <- function(project) {\n",
    "  filename <- paste0(\n",
    "    \"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/code/09_drug_repository/09_4_drexml_gtex/\",\n",
    "    tolower(paste(proj_tumorvsctrl[proj_tumorvsctrl$tcga == project,\"gtex\"],project,sep=\"_\")),\n",
    "    \"/launcher_custom_circuits_female.sh\"\n",
    "  )\n",
    "  \n",
    "  script <- paste(\n",
    "    \"#!/bin/bash\\n\",\n",
    "    \"#SBATCH -J dr-schizo-launcher\\n\",\n",
    "    \"#SBATCH -t 0\\n\",\n",
    "    \"#SBATCH --mem 8G\\n\",\n",
    "    \"#SBATCH -n 1\\n\",\n",
    "    \"#SBATCH -o pipeline_custom_circuits.log\\n\\n\",\n",
    "    \"###################################################\\n\",\n",
    "    \"# SET THE FOLLOWING VARIABLES TO MATCH YOUR NEEDS #\\n\",\n",
    "    \"###################################################\\n\",\n",
    "    \"SINGULARITY_MOD='Singularity/3.9.7'\\n\",\n",
    "    \"MAMBA_MOD='Mamba/4.14.0-0'\\n\",\n",
    "    \"PARAMS_FILE='/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/code/09_drug_repository/09_4_drexml_gtex/\",\n",
    "    tolower(paste(proj_tumorvsctrl[proj_tumorvsctrl$tcga == project,\"gtex\"],project,sep=\"_\")),\n",
    "    \"/params_custom_circuits_female.yml'\\n\",\n",
    "    \"CONDA_INIT_SCRIPT='/mnt/lustre/expanse/software/easybuild/Rocky/8.5/Skylake/software/Mamba/4.14.0-0/etc/profile.d/conda.sh'\\n\",\n",
    "    \"CONDA_ENV='nextflow'\\n\",\n",
    "    \"UPDATE_NEXTFLOW='TRUE'\\n\\n\",\n",
    "    \"# The PIPELINE_REPO variable must point to the github repository (user/project) or the main.nf full path (in your local -cloned- repository)\\n\",\n",
    "    \"#PIPELINE_REPO='babelomics/dr-pipeline'\\n\",\n",
    "    \"PIPELINE_REPO='/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/code/09_drug_repository/dr-pipeline/main.nf'\\n\\n\",\n",
    "    \"########\\n\",\n",
    "    \"# INIT #\\n\",\n",
    "    \"########\\n\",\n",
    "    \"# check params file\\n\",\n",
    "    \"[[ -f \\\"${PARAMS_FILE}\\\" ]] || { echo \\\"Params file ${PARAMS_FILE} not found!\\\" && exit 1; }\\n\\n\",\n",
    "    \"# check if singularity module is loaded\\n\",\n",
    "    \"[[ $(module is-loaded ${SINGULARITY_MOD}) ]] || module load ${SINGULARITY_MOD}\\n\\n\",\n",
    "    \"# check if mamba module is loaded\\n\",\n",
    "    \"[[ $(module is-loaded ${MAMBA_MOD}) ]] || module load ${MAMBA_MOD}\\n\\n\",\n",
    "    \"# initialize conda\\n\",\n",
    "    \"source ${CONDA_INIT_SCRIPT} || { echo \\\"The conda environment couldn't be initialized!\\\" && exit 1; }\\n\\n\",\n",
    "    \"# check conda env\\n\",\n",
    "    \"if [[ ! $(conda info --envs |grep ${CONDA_ENV}) ]]\\n\",\n",
    "    \"then\\n\",\n",
    "    \"  echo \\\"${CONDA_ENV} environment not found. Trying to create a new conda environment...\\\"\\n\",\n",
    "    \"  conda create -y -n ${CONDA_ENV} bioconda::nextflow\\n\",\n",
    "    \"fi\\n\\n\",\n",
    "    \"# activate the conda environment\\n\",\n",
    "    \"conda activate ${CONDA_ENV} || { echo \\\"The conda environment couldn't be activated!\\\" && exit 1; }\\n\\n\",\n",
    "    \"# update nextflow\\n\",\n",
    "    \"[[ \\\"${UPDATE_NEXTFLOW}\\\" ]] && nextflow self-update\\n\\n\",\n",
    "    \"########\\n\",\n",
    "    \"# MAIN #\\n\",\n",
    "    \"########\\n\",\n",
    "    \"# run the workflow from the github repository\\n\",\n",
    "    \"nextflow run ${PIPELINE_REPO} -profile slurm,singularity -resume -params-file ${PARAMS_FILE}\\n\",\n",
    "    sep=\"\"\n",
    "  )\n",
    "  \n",
    "  write(script, file = filename)\n",
    "}\n",
    "\n",
    "\n",
    "lapply(proj_tumorvsctrl$tcga, FUN=function(x) create_file_launcher_custom_circuits_female(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "e7317466-5171-4030-9ff3-4d5973a4ee22",
   "metadata": {},
   "outputs": [],
   "source": [
    "writeLines(unlist(lapply(tolower(paste(proj_tumorvsctrl$gtex, proj_tumorvsctrl$tcga, sep=\"_\")), FUN=function(x) \n",
    "paste(\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/code/09_drug_repository/09_4_drexml_gtex/\",x,\"/launcher_custom_circuits_male.sh\",sep=\"\"))),\n",
    "\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/code/09_drug_repository/09_4_drexml_gtex/09_4_drexml_gtex_male.cmd\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "5083b838-4bd7-419b-abd4-077bde920b7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "writeLines(unlist(lapply(tolower(paste(proj_tumorvsctrl$gtex, proj_tumorvsctrl$tcga, sep=\"_\")), FUN=function(x) \n",
    "paste(\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/code/09_drug_repository/09_4_drexml_gtex/\",x,\"/launcher_custom_circuits_female.sh\",sep=\"\"))),\n",
    "\"/mnt/lustre/scratch/CBRA/research/projects/tcga_mm/code/09_drug_repository/09_4_drexml_gtex/09_4_drexml_gtex_female.cmd\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af54323f-0e8a-4bbe-99ea-d01a7582cca6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
